lib/puma/server.rb

```diff
@@ -330,8 +331,11 @@ def handle_servers
 
         while @status == :run || (drain && shutting_down?)
           begin
-            ios = IO.select sockets, nil, nil, (shutting_down? ? 0 : nil)
-            break unless ios
+            ios = IO.select sockets, nil, nil, (shutting_down? ? 0 : timeout_for_idle)
+            unless ios
+              @status = :stop if timeout_for_idle&.zero?
```

Comment:
This is a super-hot loop... I need to think about this.

Reply:
I'm quite unfamiliar with Puma's codebase, but why is system time is even needed here?

The feature is to exit if nothing was read in X seconds? Why not just make that the select timeout and exit if `IO.select` doesn't return any IO?

Reply:
> I'm quite unfamiliar with Puma's codebase

Same here, so I'm hoping someone more experienced will correct me if I'm wrong.

> Why not just make that the select timeout and exit if IO.select doesn't return any IO?

When there's no timeout and `IO.select` simply doesn't return any IO, the server doesn't shut down gracefully (at least from what I found in my tests). So not only do we know that there's no IO, we also need to know that the timeout has elapsed to gracefully shut it down explicitly, which we need to do if we want to keep the sockets received from systemd open.

Reply:
> When there's no timeout and `IO.select` simply doesn't return any IO

I don't follow. If you don't pass a timeout, `IO.select` will block forever.

So if `select` returns `nil`, you are either in the `shutting_down?` case, or in the `idle_timeout?` case. Or am I missing something?

Reply:
> So if `select` returns `nil`, you are either in the `shutting_down?` case, or in the `idle_timeout?` case. Or am I missing something?

That's what I thought too, but I found that it isn't `shutting_down?` when `select` returns `nil`. But by setting the status to `:stop`, I'm able to enter the shutdown phase due to:
https://github.com/puma/puma/blob/2563396f730adef54a9369dbae996fda3bba5f6e/lib/puma/server.rb#L365

Reply:
You mean `IO.select sockets, nil, nil, nil` returns `nil`? 

It's really not supposed to happen.

Reply:
Ya okay I was definitely wrong about that. I think I was just trying to not break this previous logic:

```ruby
break unless ios
```

But it sounds like that case shouldn't even be possible unless it times out?
All my tests pass if I just use your suggestion AND set the status to `:stop` if `ios` is `nil`. Which makes sense to me. 

However, I do have this one failing case:

https://github.com/joshuay03/puma/blob/23fc9b2d1c49a64ebdd9f36dcd8c9c22d67d5bc1/test/test_puma_server.rb#L632-L644

Which also makes sense cause I approached this as `"the timeout since the last request"` and not `"the timeout until the next request"`. But if that expectation isn't necessary, I think this works perfectly!

Reply:
> So I think you can just check shutting_down? after the unless ios to know if you reached the idle timeout or are simply shutting down already.

Perfect! Thank you so much!

Reply:
> Which also makes sense cause I approached this as `"the timeout since the last request"` and not `"the timeout until the next request"`. But if that expectation isn't necessary, I think this works perfectly!

Actually, I think I can figure out a way to handle this, will report back!

Reply:
Resolved with: 9c4a15a158b554f67e59f33378eee10abeec8c14.